components/ai_chat/resources/untrusted_conversation_frame/components/markdown_renderer/index.tsx

```diff
@@ -252,61 +247,249 @@ function buildTableRenderer() {
 
 interface MarkdownRendererProps {
   text: string
-  shouldShowTextCursor: boolean
+  isEntryInProgress: boolean
   allowedLinks?: string[]
   disableLinkRestrictions?: boolean
+  elementTextStateRef?: React.MutableRefObject<Map<number, string>>
 }
 
-export default function MarkdownRenderer(mainProps: MarkdownRendererProps) {
-  const lastElementRef = React.useRef<HastElement | undefined>()
+/**
+ * Extracts all text content from a Hast (HTML AST) node recursively.
+ * Used for tracking text length changes in elements for fade-in animation.
+ */
+function extractTextFromNode(node: any): string {
```

Comment:
nit: should this be typed as a Node?

```suggestion
function extractTextFromNode(node: Node): string {
```